mirascope/integrations/wandb/wandb_prompt.py

```diff
@@ -0,0 +1,90 @@
+
+from pydantic import BaseModel
+from wandb.sdk.data_types.trace_tree import Trace
+
+from mirascope import BaseCallParams, Prompt
+from mirascope.chat import OpenAIChatCompletion
+
+
+class WandbPrompt(Prompt):
+    """Parent class for inherited WandB functionality."""
+
+    span_type: Literal["tool", "llm"]
+
+    call_params = BaseCallParams(model="gpt-3.5-turbo-0125")
+
+    def trace(
+        self, completion: Union[OpenAIChatCompletion, BaseModel], parent: Trace
+    ) -> Trace:
+        """Returns a trace connected to parent.
+
+        Args:
+            completion: The completion to trace. Handles `OpenAIChatCompletion` output
+                from both standard OpenAI chat completions, and `BaseModel` for
+                extractions.
+            parent: The parent trace to connect to.
+
+        Returns:
+            The created trace, connected to the parent.
+        """
+        if isinstance(completion, OpenAIChatCompletion):
+            if completion.tool and completion.tool.fn:
+                output = {
+                    "assistant": completion.tool,
+                    "tool_output": completion.tool.fn(
+                        **completion.tool.model_dump(exclude={"tool_call"})
+                    ),
+                }
+            else:
+                output = {"assistant": str(completion)}
+            dump = completion.dump()
+        elif isinstance(completion, BaseModel):
+            output = {"assistant": str(completion.model_dump())}
+            if not hasattr(completion, "_completion"):
+                raise ValueError(
+                    "Completion of type `BaseModel` was not created using the `extract`"
+                    " function and does not contain the necessary `_completion` private"
+                    " attribute."
+                )
+            else:
+                dump = completion._completion.dump()
+
+        span = Trace(
+            name=self.__class__.__name__,
+            kind=self.span_type,
+            status_code="success",
+            status_message=None,
+            metadata={"model": self.call_params.model},
+            start_time_ms=dump["start_time"],
+            end_time_ms=dump["end_time"],
+            inputs={message["role"]: message["content"] for message in self.messages},
+            outputs=output,
+        )
+        parent.add_child(span)
+        return span
+
+    def trace_error(self, error: Exception, parent: Trace) -> Trace:
+        """Returns an error trace connected to parent.
+
+        Args:
+            error: The error to trace.
+            parent: The parent trace to connect to.
+
+        Returns:
+            The created error trace, connected to the parent.
+        """
+        span = Trace(
+            name=self.__class__.__name__,
+            kind=self.span_type,
+            status_code="error",
+            status_message=str(error),
+            metadata={"model": self.call_params.model},
+            start_time_ms=None,
```

Comment:
do errors not care about start time end time?

Reply:
Interesting, there's no completion to retrieve the start/end time conveniently, but we do likely want it

Reply:
if we want it for errors, we would have to call get_time of some form before every call to a chat completion, since calculating it on the fly only gets the end time. do we want to implement start time and end time as optional parameters in the trace_error so that they users can pass it in if they choose to?

Reply:
What if the error threw the start time @brenkao ?

Reply:
Decision: generate start time on prompt creation and end time on trace_error call.